mopidy/files/library.py

```diff
@@ -0,0 +1,157 @@
+            for name in os.listdir(localpath):
+                child = os.path.join(localpath, name)
+                uri = path.path_to_uri(child)
+                name = name.decode('ascii', 'ignore')
+                if self._follow_symlinks:
+                    st = os.stat(child)
+                else:
+                    st = os.lstat(child)
+                if not self._show_hidden and name.startswith(b'.'):
+                    continue
+                elif stat.S_ISDIR(st.st_mode):
+                    result.append(models.Ref.directory(name=name, uri=uri))
+                elif stat.S_ISREG(st.st_mode) and self._check_audiofile(uri):
+                    result.append(models.Ref.track(name=name, uri=uri))
+                else:
+                    logger.warn(u'Ignored file: %s' % child.decode('ascii',
+                                                                   'ignore'))
+                    continue
+
+        result.sort(key=operator.attrgetter('name'))
+        return result
+
+    def lookup(self, uri):
+        logger.debug(u'looking up uri = %s' % uri)
+        localpath = path.uri_to_path(uri)
+        if not self._is_in_basedir(localpath):
+            logger.warn(u'Not in basedir: %s' % localpath)
+            return []
+        try:
+            result = self._scanner.scan(uri)
+            track = utils.convert_tags_to_track(result.tags).copy(
+                uri=uri, length=result.duration)
+        except exceptions.ScannerError as e:
+            logger.warning(u'Problem looking up %s: %s', uri, e)
+            track = models.Track(uri=uri)
+            pass
+        if not track.name:
+            filename = os.path.basename(localpath)
+            name = urllib2.unquote(filename).decode('ascii', 'ignore')
+            track = track.copy(name=name)
+        return [track]
+
+    def _show_media_dirs(self):
+        result = []
+        for media_dir in self._media_dirs:
+            dir = models.Ref.directory(
+                name=media_dir['name'],
+                uri=path.path_to_uri(media_dir['path']))
+            result.append(dir)
+        return result
+
+    def _check_audiofile(self, uri):
+        try:
+            result = self._scanner.scan(uri)
+            logger.debug(u'got scan result playable: %s for %s' %
+                         (result.uri, str(result.playable)))
+            res = result.playable
+        except exceptions.ScannerError as e:
+            logger.warning(u'Problem scanning %s: %s', uri, e)
+            res = False
+        return res
+
+    def _is_playlist(self, child):
+        return os.path.splitext(child)[1] == '.m3u'
+
+    def _is_in_basedir(self, localpath):
+        res = False
+        basedirs = [mdir['path'] for mdir in self._media_dirs]
+        for basedir in basedirs:
+            if basedir == localpath:
+                res = True
+            else:
+                try:
+                    path.check_file_path_is_inside_base_dir(localpath, basedir)
```

Comment:
Ah, i see. The function is written with only files in mind. (as the name suggests)
I am also checking directories. When checking on directories
    real_dir_path = os.path.dirname(real_file_path)
is not necessary.
I could now:
1. check only on files. This means the user could browse directories outside the media directories, but no files will be listed.
2. rewrite the function to check directories and files.
3. write a new function to check directories, put a wrapper method in front and place common code in a helper method. This would make sense if other parts of mopidy will ever use this methods.

i will go with option 1 for now, as it is the least intrusive.